### PR TITLE
[support spark2.3] importCode from spark2.3 diff request

### PR DESCRIPTION
**Pull request by zhixingheyi-tian**
Changes made to the imported spark 2.3 source code

Among these source codes。
ColumnarBatch.java  、FileFormatWriter.scala  varies greatly from spark 2.2 to spark 2.3.
Please have more reviews on "ColumnarBatch.java 、FileFormatWriter.scala" in the OAP project.

This is a split "pull request", unit tests passed in the following PR
#850 

Thanks!

### CODE REVIEW

**Comment by gczsjdy** [comment]
Again, please use the template and fill in `How was this patch tested?`

File: src/main/spark2.3/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -307,6 +307,31 @@ class DAGScheduler(
...
+      }
+      if (oapPrefs.nonEmpty) {
+        logDebug(s"got oap prefer location value oapPrefs is ${oapPrefs}")
+        return oapPrefs.map(loc => TaskLocation(loc._1, loc._2))
+      }
+    }
+    Seq.empty
+  }
+
+
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/scheduler/DAGScheduler.scala]: duplicated empty line

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala [hunk 2]
```diff
@@ -1,171 +1,172 @@
...
-       |while ($input.hasNext()) {
-       |  InternalRow $row = (InternalRow) $input.next();
-       |  $numOutputRows.add(1);
-       |  ${consume(ctx, outputVars, inputRow).trim}
-       |  if (shouldStop()) return;
-       |}
-     """.stripMargin
-  }
-}
+/*
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: What change for ColumnBatchScan?

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 3]
```diff
@@ -85,7 +85,9 @@ object FileFormatWriter extends Logging {
   }
 
   /** The result of a successful write task. */
-  private case class WriteTaskResult(commitMsg: TaskCommitMessage, summary: ExecutedWriteSummary)
+  private case class WriteTaskResult(commitMsg: TaskCommitMessage,
+                                     summary: ExecutedWriteSummary,
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: spark code style

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala (hunk 2 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: @JkSelf 
On line 138, the following line is added.
|    if ($batch.isFiltered($rowidx)) continue;

File: src/main/spark2.3/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Have been adjusted.
Thanks!

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 3 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: @JkSelf 
Have been adjusted.
Thanks!

**Comment by zhixingheyi-tian** [comment]
@gczsjdy 
About "tests", now updated in the description

**Comment by zhixingheyi-tian** [comment]
@JkSelf 
It has now been commited for  this PR.
Generally, the main PR is updated first. The split PR needs to switch branches to re-operate. Sometimes there will be a delay.

Thanks.

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java [hunk 4]
```diff
@@ -75,6 +86,42 @@ public void remove() {
     };
   }
 
+  /**
+   * Returns an iterator over the rows in this batch. for OAP
+   */
+  public Iterator<InternalRow> rowOapIterator() {
```
Comment (carsonwang) [inline src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: Where is this used?

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java [hunk 5]
```diff
@@ -35,6 +35,12 @@
   // Staging row returned from `getRow`.
   private final MutableColumnarRow row;
 
+  // True if the row is filtered.
+  private final boolean[] filteredRows = new boolean[1 << 16];
```
Comment (carsonwang) [inline src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: Why to use 1 << 16? What is used in spark 2.2? Better to define a static variable.

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala (hunk 2 above)
Comment (carsonwang) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: Did you import a wrong file earlier?

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 6]
```diff
@@ -1626,4 +1627,86 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
...
+        Option(ctx.partitionSpec).map(visitNonOptionalPartitionSpec))
+    }
+
+  override def visitOapDisableIndex(ctx: OapDisableIndexContext): LogicalPlan =
+    OapDisableIndexCommand(ctx.IDENTIFIER.getText)
+
+  override def visitOapEnableIndex(ctx: OapEnableIndexContext): LogicalPlan =
+    OapEnableIndexCommand(ctx.IDENTIFIER.getText)
+
+
```
Comment (carsonwang) [inline src/main/spark2.3/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Remove above two empty lines.

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -524,7 +525,7 @@ case class DataSource(
               s"Unable to resolve $name given [${data.output.map(_.name).mkString(", ")}]")
           }
         }
-        val resolved: InsertIntoHadoopFsRelationCommand = cmd.copy(partitionColumns = resolvedPartCols, outputColumns = outputColumns)
+        val resolved = cmd.copy(partitionColumns = resolvedPartCols, outputColumns = outputColumns)
```
Comment (carsonwang) [inline src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Why did you change this? This is Spark's original code and let's keep it.

Comment (carsonwang) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Did you import a wrong file earlier?

File: src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+import org.apache.spark.unsafe.Platform;
+
+import org.apache.parquet.column.values.ValuesReader;
+import org.apache.parquet.io.api.Binary;
+
+/**
+ * An implementation of the Parquet PLAIN decoder that supports the vectorized interface.
+ */
+public class VectorizedPlainValuesReader extends ValuesReader implements VectorizedValuesReader {
```
Comment (carsonwang) [inline src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java]: can you please explain why to import this file and what is changed?

File: src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+ * A values reader for Parquet's run-length encoded data. This is based off of the version in
+ * parquet-mr with these changes:
+ *  - Supports the vectorized interface.
+ *  - Works on byte arrays(byte[]) instead of making byte streams.
+ *
+ * This encoding is used in multiple places:
+ *  - Definition/Repetition levels
+ *  - Dictionary ids.
+ */
+public class VectorizedRleValuesReader extends ValuesReader
```
Comment (carsonwang) [inline src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: can you please explain why to import this file and what is changed?

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala (hunk 2 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: @carsonwang 
After checking, the originally imported ColumnarBatchScan.scala has newline characters  from the dos system. So a lot of invalid diffs are generated.
In fact, the difference is only this line.

Thanks

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java (hunk 4 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: @carsonwang 

rowOapIterator() is used by ParquetDataFile.scala, ParquetDataFileSuite.scala.
In the version of spark2.1 spark2.2, ParquetDataFile.scala and ParquetDataFileSuite.scala use rowIterator() , but rowIterator() removes the filteredRows related functionality in the version of spark2.3. So added this method.

Thanks.

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java (hunk 5 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: @carsonwang 

The properties and methods related to filteredRows have been removed in the version of spark2.3. This member variable was added to support OAP related functions.
  ”1 << 16” is the default maximum capacity value.
In the version of spark2.1, spark2., this value is passed in by related methods. In the version of spark2.3, these methods were removed or changed too large to support incoming parameters.

I once considered defining a static variable. But did not find a suitable position for this variable.

Thanks.

**Comment by zhixingheyi-tian** [comment]
@carsonwang 

The following code is added because linhong merged the relevant code a few days ago. Some of his code needs to distinguish between different versions.

VectorizedPlainValuesReader.java
VectorizedRleValuesReader.java

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 7 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: @carsonwang 

After checking, the original code of spark does not have ": InsertIntoHadoopFsRelationCommand". It seems that when the code is imported into IDEA, the related button is accidentally touched, and the type annotation is automatically added.

Thanks

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java (hunk 5 above)
Comment (carsonwang) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: Can you take a look at 2.1 and 2.2, it defines DEFAULT_BATCH_SIZE = 4 * 1024

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 6 above)
Comment (carsonwang) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: I didn't see this was resolved but was marked as resolved.

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java (hunk 4 above)
Comment (carsonwang) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: Please see https://github.com/apache/spark/pull/19766 why the filtered rows are removed.

Comment (carsonwang) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: can you remove rowOapIterator and still update rowIterator? It is ok as Spark doesn't use the filtered row feature. Then there is no need to create different versions for ParquetDataFile.scala, ParquetDataFileSuite.scala.

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 10]
```diff
@@ -52,7 +55,7 @@ import org.apache.spark.sql.execution.SparkPlan
 object FileSourceStrategy extends Strategy with Logging {
   def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
     case PhysicalOperation(projects, filters,
-      l @ LogicalRelation(fsRelation: HadoopFsRelation, _, table, _)) =>
+      l @ LogicalRelation(_fsRelation: HadoopFsRelation, _, table, _)) =>
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Why change the value from fsRelation to _fsRelation?

File: src/main/spark2.3/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala [hunk 11]
```diff
@@ -51,6 +54,43 @@ private[v1] class AbstractApplicationResource extends BaseAppResource {
   @Path("executors")
   def executorList(): Seq[ExecutorSummary] = withUI(_.store.executorList(true))
 
+  @GET
+  @Path("fibercachemanagers")
+  def fiberList(): Seq[FiberCacheManagerSummary] =
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala]: What is the purpose of this method?

File: src/main/spark2.3/scala/org/apache/spark/util/collection/BitSet.scala [hunk 12]
```diff
@@ -34,6 +34,16 @@ class BitSet(numBits: Int) extends Serializable {
...
 
+  def toLongArray(): Array[Long] = words
+
+  def this(from: Array[Long]) {
+    this(from.length << 6)
+    for(i <- from.indices) {
+      words(i) = from(i)
+    }
+  }
+
```
Comment (JkSelf) [inline src/main/spark2.3/scala/org/apache/spark/util/collection/BitSet.scala]: remove repeated empty line?

File: src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java (hunk 4 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java]: @carsonwang 
As discussed offline, it is more appropriate to retain "rowOapIterator".

File: src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java (hunk 8 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java]: @carsonwang

The following code is added because linhong merged the relevant code a few days ago. Some of his code needs to distinguish between different versions.

VectorizedPlainValuesReader.java
VectorizedRleValuesReader.java

File: src/main/spark2.3/scala/org/apache/spark/util/collection/BitSet.scala (hunk 12 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/util/collection/BitSet.scala]: @JkSelf 
Has been resolved. 

Thanks

File: src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 10 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: @JkSelf 

If you look down a few lines, there's the variable "fsRelation" below, which is also consistent with Spark 2.1 and Spark 2.2

Thanks

File: src/main/spark2.3/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala (hunk 11 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala]: @JkSelf 

For OAP  UI.

Thanks

File: src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java (hunk 9 above)
Comment (zhixingheyi-tian) [reply src/main/spark2.3/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java]: @carsonwang 

The reason is the same as above.

Thanks

**Comment by carsonwang** [comment]
Thanks
